jmdaemon/jmdaemon/onionmc.py

```diff
@@ -0,0 +1,1214 @@
+        # connected (note we could use a deferred but
+        # the rpc connection calls are not using twisted)
+        self.wait_for_directories_loop = None
+
+    def info_callback(self, msg: str) -> None:
+        log.info(msg)
+
+    def setup_error_callback(self, msg: str) -> None:
+        log.error(msg)
+
+    def shutdown_callback(self, msg: str) -> None:
+        log.info("in shutdown callback: {}".format(msg))
+
+    def onion_hostname_callback(self, hostname: str) -> None:
+        """ This entrypoint marks the start of the OnionMessageChannel
+        running, since we need this unique identifier as our name
+        before we can start working (we need to compare it with the
+        configured directory nodes).
+        """
+        log.info("setting onion hostname to : {}".format(hostname))
+        self.onion_hostname = hostname
+
+# ABC implementation section
+    def run(self) -> None:
+        self.hs_up_loop = task.LoopingCall(self.check_onion_hostname)
+        self.hs_up_loop.start(0.5)
+
+    def get_pubmsg(self, msg:str, source_nick:str ="") -> str:
+        """ Converts a message into the known format for
+        pubmsgs; if we are not sending this (because we
+        are a directory, forwarding it), `source_nick` must be set.
+        Note that pubmsg does NOT prefix the *message* with COMMAND_PREFIX.
+        """
+        nick = source_nick if source_nick else self.nick
+        return nick + COMMAND_PREFIX + "PUBLIC" + msg
+ 
+    def get_privmsg(self, nick: str, cmd: str, message: str,
+                    source_nick=None) -> str:
+        """ See `get_pubmsg` for comment on `source_nick`.
+        """
+        from_nick = source_nick if source_nick else self.nick
+        return from_nick + COMMAND_PREFIX + nick + COMMAND_PREFIX + \
+               cmd + " " + message
+
+    def _pubmsg(self, msg:str) -> None:
+        """ Best effort broadcast of message `msg`:
+        send the message to every known directory node,
+        with the PUBLIC message type and nick.
+        """
+        dps = self.get_directory_peers()
+        msg = OnionCustomMessage(self.get_pubmsg(msg),
+                                JM_MESSAGE_TYPES["pubmsg"])
+        for dp in dps:
+            # currently a directory node can send its own
+            # pubmsgs (act as maker or taker); this will
+            # probably be removed but is useful in testing:
+            if dp == self.self_as_peer:
+                self.receive_msg(msg, "00")
+            else:
+                self._send(dp, msg)
+
+    def _privmsg(self, nick: str, cmd: str, msg:str) -> None:
+        # in certain test scenarios the directory may try to transfer
+        # commitments to itself:
+        if nick == self.nick:
+            log.debug("Not sending message to ourselves: {}, {}, {}".format(
+                nick, cmd, msg))
+            return
+        log.debug("Privmsging {}, {} to: {}".format(cmd, msg, nick))

```

Comment:
Noticed while testing that this is duplicating the debug message.
E.g.,
```bash
2022-03-19 11:21:09,219 [DEBUG]  >>privmsg on onion-network: nick=J57KRpr1hfYSRa3z cmd=fill msg=0 200000 cf7f90e0100c839dd58453b6a9da08d873763d9dfdc53ddc73baf220d58c345e Pe44a182de636339e835deffdf549498da6ece68c9cf4935e657c7dfc7d2e0592 03422a9e6ead10064ffb09f9ebd4b22485beed691c3df4ae2738bedba46033933f MEUCIQDCcwMMMDvermhj/PvjuXnpHZB1pdsiWBRBtQsU+sMn5gIgTQUa3aGaD1DHHhCDSVDYwMJ7LUXvvQ1slkr2YnVTc2A=
2022-03-19 11:21:09,220 [DEBUG]  Privmsging fill, 0 200000 cf7f90e0100c839dd58453b6a9da08d873763d9dfdc53ddc73baf220d58c345e Pe44a182de636339e835deffdf549498da6ece68c9cf4935e657c7dfc7d2e0592 03422a9e6ead10064ffb09f9ebd4b22485beed691c3df4ae2738bedba46033933f MEUCIQDCcwMMMDvermhj/PvjuXnpHZB1pdsiWBRBtQsU+sMn5gIgTQUa3aGaD1DHHhCDSVDYwMJ7LUXvvQ1slkr2YnVTc2A= to: J57KRpr1hfYSRa3z
```

Reply:
Yes; this is a bit annoying indeed. Hopefully will find a moment to fix it at some point.

Reply:
OK, was a holdover from testing (making sure the privmsg function works). Duplicate removed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/1182/commits/5fb15ced3608b70d55e40d70ede0f271673fd4ac